frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -801,10 +801,15 @@ func toDispatchState(ctx context.Context, dt []byte, opt ConvertOpt) (*dispatchS
 	target.state = target.state.SetMarshalDefaults(defaults...)
 
 	if !platformOpt.implicitTarget {
+		sameOsArch := platformOpt.targetPlatform.OS == target.image.OS && platformOpt.targetPlatform.Architecture == target.image.Architecture
 		target.image.OS = platformOpt.targetPlatform.OS
 		target.image.Architecture = platformOpt.targetPlatform.Architecture
-		target.image.Variant = platformOpt.targetPlatform.Variant
-		target.image.OSVersion = platformOpt.targetPlatform.OSVersion
+		if platformOpt.targetPlatform.Variant != "" || !sameOsArch {
+			target.image.Variant = platformOpt.targetPlatform.Variant
+		}
+		if platformOpt.targetPlatform.OSVersion != "" || !sameOsArch {
+			target.image.OSVersion = platformOpt.targetPlatform.OSVersion
+		}
 		if platformOpt.targetPlatform.OSFeatures != nil {
```

Comment:
This `OSFeatures` line doesn't make much sense to me either as it can create duplicates and if `!sameOsArch` then probably should not be merged. Shouldn't be impactful as `OSFeatures` can't be passed via `opt` atm. but that will change. I opened https://github.com/containerd/platforms/issues/23 as I think some of it should be handled in upstream platforms pkg.

Reply:
This isn't merging the OSFeatures, it's replacing them, so I don't understand where duplicates would come from? It should just be a straight-up copy of `platformOpt.targetPlatform.OSFeatures`.

It probably should have had the same `!sameOsArch` check added though. Not particularly impactful, last time I looked into it, `OSFeatures` was 100% unused and the OCI Working Group that might want it for future use didn't seem interested in it, in favour of a separate artifact that covered its use-cases.

Reply:
> This isn't merging the OSFeatures, it's replacing them

Ah, yeah. I guess I looked that it was appending to the old `OSFeatures`, but it is appending to empty array (`slices.Clone`)

Reply:
Yeah, I had to open-code `slices.Clone` in the original PR because it wasn't available until Go 1.21 and we were declaring Go 1.20 as supported at that time.

That's no longer the case so we could clean that up for readability now.